Azure.Search: Add durable continuation tokens

tg-msft opened the pull request:
Fixes #10590

The way continuation tokens are used is slightly odd, but it's necessary given how we've hidden the paging behind a request.  We also hope that most customers won't even be aware of server-side paging in Track 2 and everything will "just work" for them.

tg-msft committed:
Add Search continuation tokens

Fixes #10590

The way continuation tokens are used is slightly odd, but it's necessary given
how we've hidden the paging behind a request.

tg-msft commented:
//fyi @xirzec / @sima-zhu / @alzimmermsft / @xiangyan99 

All of the logic for token serialization is in
https://github.com/Azure/azure-sdk-for-net/blob/f8fc68dd4af4d6136e979a8cd388e7904a54f482/sdk/search/Azure.Search.Documents/src/Models/SearchContinuationToken.cs if you want to have a look for consistency's sake.

tg-msft committed:
Update the exported API

heaths commented on sdk/search/Azure.Search.Documents/src/Models/SearchContinuationToken.cs:
```diff
@@ -0,0 +1,104 @@
+            {
+                using JsonDocument json = JsonDocument.Parse(decoded);
+                if (json.RootElement.ValueKind == JsonValueKind.Object &&
+                    json.RootElement.TryGetProperty(ApiVersionName, out JsonElement apiVersion) &&
```
`TryGetProperty` is slower. Each call is O(n) where n is the number of child elements. Instead, follow the same pattern as the generated serialize and use `EnumerateObject()`.

pakrym replied:
I don't quite understand this, EnumerateObject does N checks N times so it's same O(N*N) as `TryGetProperty` or am I missing something?

heaths replied:
No, it loops through each property. You'd use a switch or if condition to handle each property name. You do this in your generator, and we do it in our other non-generated services already. Each call to TryGetProperty is O(n), so three calls triples the time (still technically O(n), but big O notation isn't an absolute order of perf). If you need to get more than one property, you use `EnumerateObject`.

pakrym replied:
I still don't understand.

```
foreach (prop in EnumerateObject())
{
	if (prop.name == "a") {}
	if (prop.name == "b") {}
	if (prop.name == "c") {}
}
```

is still a N*N, as it does N loops with N comparisons in each


Same as 
```
TryGetProperty("a")
TryGetProperty("b")
TryGetProperty("c")
```
where every call does 1 loops with N comparisons in each.

heaths replied:
The first is O(n). If you want to count the comparison, it's O(nm). The relative performance of the latter would be O(3nm). The latter also allocates (even though on the stack, it's still an extra allocation) for each enumerate it creates internally. We've discussed this as a team previously and prefer `EnumerateObject`. You're doing it in all the deserialization code right now. Why is this an issue all of a sudden?

tg-msft replied:
Given that it will only have 3 properties, I'd prefer to leave it written like this because I think it's cleaner than turning it into a state machine and O(27) doesn't bother me for niche scenarios we don't really want customers using.  I will also be adding an additional `ValueKind` check here after pinging @ahsonkhan about it (the `TryGet*` methods will still throw if the `ValueKind` is incorrect...  I didn't grasp that initially).

I can change it to use `EnumerateObject` if everyone thinks this a poor choice.

pakrym replied:
It's not an issue, I'm fine with both implementations I just don't understand the reasoning and I want to.

pakrym replied:
TryGetProperty seems to be faster:

```

|         Method |     Mean |    Error |   StdDev |
|--------------- |---------:|---------:|---------:|
| TryGetProperty | 499.9 ns | 13.33 ns | 39.30 ns |
|           Loop | 738.6 ns | 22.75 ns | 67.08 ns |
```

```
using System;
using System.Collections.Generic;
using System.Diagnostics;
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;
using System.Text.Json;

namespace activity_bench
{
    public class JElementBench
    {
        public static readonly JsonElement e = JsonDocument.Parse(@"{
           ""a"":""a"",
           ""b"":""b"",
           ""c"":""c"",
           ""d"":""d"",
           ""f"":""f""
            }").RootElement;

        [Benchmark]
        public void TryGetProperty()
        {
            e.TryGetProperty("a", out var a);
            e.TryGetProperty("b", out var b);
            e.TryGetProperty("c", out var c);
            e.TryGetProperty("d", out var d);
            e.TryGetProperty("f", out var f);
        }

        [Benchmark]
        public void Loop()
        {
            JsonElement a;
            JsonElement b;
            JsonElement c;
            JsonElement d;
            JsonElement f;

            foreach (var prop in e.EnumerateObject())
            {
                if (prop.Name == "a") a = prop.Value;
                if (prop.Name == "b") a = prop.Value;
                if (prop.Name == "c") a = prop.Value;
                if (prop.Name == "d") a = prop.Value;
                if (prop.Name == "f") a = prop.Value;
            }
        }
    }

    class Program
    {
        static void Main(string[] args)
        {
            var summary = BenchmarkRunner.Run<JElementBench>();
        }
    }
}

```

heaths commented on sdk/search/Azure.Search.Documents/src/Models/SearchContinuationToken.cs:
```diff
@@ -0,0 +1,104 @@
+            {
+            }
+
+            throw new ArgumentException("Invalid continuation token", nameof(continuationToken));
```
Given this whole class is internal, I think it's confusing to throw an `ArgumentException`. Besides, allowing the exception above to throw out is both consistent with all our other deserialization code and provides context about what exactly is wrong. And because the parameter name here is inconsequential to the user anyway, it's certainly *no worse* - only better because it also has more information about what exactly is wrong.

tg-msft replied:
The parameter name `continuationToken` is used by all of its (one) caller(s).  It felt cleaner to throw ArgEx here than throw for some conditions/return null and then catch/rewrap in ArgEx at the call site.  We also want to treat these continuationTokens opaquely (for the moment, at least) so I don't want to get specific that certain fields weren't found, etc.  We're going to revisit this in the next preview for consistency across languages.

tg-msft committed:
Adding an extra ValueKind check based on feedback from Ahson

and removing apiVersion param since it should always be a constant at this point

heaths commented on sdk/search/Azure.Search.Documents/src/Models/SearchResults{T}.cs:
```diff
@@ -231,7 +229,12 @@ public SearchAsyncPageable(SearchResults<T> results)
+            // The first page of our results is always provided so we can
+            // ignore the continuation token.  Users can only provide a token
+            // directly to the Search method.
+            Debug.Assert(continuationToken == null);
```
Given the comment, are we not planning on testing this eventually then, for which this will assert?

tg-msft replied:
There's not really anything to test.  We just ignore the value.  I added a `Debug.Assert` as a friendly reminder that somebody screwed something up if this happens while we're working on it.  We could consider strengthening that to an exception but I kind of want Krzysztof to give this whole area another look anyway.   We made the call to squish Pageable behind a Response and this is a consequence.  I think it's okay, but more eyes before GA will be good.

heaths commented on sdk/search/Azure.Search.Documents/src/SearchClientOptions.cs:
```diff
@@ -38,6 +38,13 @@ public enum ServiceVersion
+        /// can be passed between different client libraries.  Changing this
+        /// value requires updating <see cref="Azure.Search.Documents.Models.SearchContinuationToken"/>.
+        /// </summary>
+        internal const ServiceVersion ContinuationTokenVersion = ServiceVersion.V2019_05_06_Preview;
```
Might we want to version this independently - maybe even monotonically (simple but effective)? Not a big deal, but seems more likely to cause some confusion if its version varies from a future service version. It's really just our own internal versioning anyway, so a monotonic version wouldn't have any unintended correlation to the api-version.

tg-msft replied:
Yep - we need a little more clarity around that.  Currently some of the languages are specifying the same version as their client but I don't think that scales well into the future.   I was thinking of using a constant service API version as a discriminator because we already have them, but `1, 2, 3, ...` would work as well.  We'll discuss it at one of the stand ups soon.

heaths commented on sdk/search/Azure.Search.Documents/tests/Utilities/TestExtensions.cs:
```diff
@@ -30,6 +30,22 @@ await foreach (T value in pageable)
+            {
+                return value;
+            }
+            Assert.Fail("No elements in sequence!");
```
Why not just throw like `First` would then? It'll still fail the text, and works more generally (i.e. not just for tests)?

tg-msft replied:
Partly because I copied existing code from above.  Partly because it's in a class called TestExtensions which I don't ever anticipate for using for more than tests.  But mostly because I copied existing code.  :smile:

heaths reviewed (approved):
A few comments, but otherwise LGTM.

tg-msft commented:
Merging to get this in today - happy to let discussion continue and I'll address any comments with an additional PR
